windows/sound.cpp

```diff
@@ -615,243 +776,107 @@ void CSound::bufferSwitch ( long index, ASIOBool )
             GetSelCHAndAddCH ( pSound->vSelectedInputChannels[i], pSound->lNumInChan,
                                iSelCH, iSelAddCH );
 
+#define CSOUND_GET_DATA(format,endian,factor)                           \
+            {                                                           \
+                format *pASIOBuf =                                      \
+                    static_cast<format *> ( pSound->bufferInfos[iSelCH].buffers[index] ); \
+                                                                        \
+                for ( iCurSample = 0; iCurSample < iASIOBufferSizeMono; iCurSample++ ) \
+                {                                                       \
+                    vecsMultChanAudioSndCrd[2 * iCurSample + i] =       \
+		        pASIOBuf[iCurSample].get##endian() * (factor);	\
+                }                                                       \
+                                                                        \
+                if ( iSelAddCH >= 0 )                                   \
+                {                                                       \
+                    /* mix input channels case: */                      \
+                    format *pASIOBufAdd =                               \
+                        static_cast<format *> ( pSound->bufferInfos[iSelAddCH].buffers[index] ); \
+                                                                        \
+                    for ( iCurSample = 0; iCurSample < iASIOBufferSizeMono; iCurSample++ ) \
+                    {                                                   \
+                        vecsMultChanAudioSndCrd[2 * iCurSample + i] =   \
+                            clipFloat ( vecsMultChanAudioSndCrd[2 * iCurSample + i] + \
+                                        pASIOBufAdd[iCurSample].get##endian() * (factor) ); \
+                    }                                                   \
+                }                                                       \
+            }
+
             // copy new captured block in thread transfer buffer (copy
             // mono data interleaved in stereo buffer)
             switch ( pSound->channelInfosInput[iSelCH].type )
             {
             case ASIOSTInt16LSB:
-            {
-                // no type conversion required, just copy operation
-                int16_t* pASIOBuf = static_cast<int16_t*> ( pSound->bufferInfos[iSelCH].buffers[index] );
-
-                for ( iCurSample = 0; iCurSample < iASIOBufferSizeMono; iCurSample++ )
-                {
-                    vecsMultChanAudioSndCrd[2 * iCurSample + i] = pASIOBuf[iCurSample];
-                }
-
-                if ( iSelAddCH >= 0 )
-                {
-                    // mix input channels case:
-                    int16_t* pASIOBufAdd = static_cast<int16_t*> ( pSound->bufferInfos[iSelAddCH].buffers[index] );
+	        CSOUND_GET_DATA(sample16,Le,1);

```

Comment:
Have you tested all possible ASIO types that they work correctly?

Reply:
No, I haven't. I've only made sure that the code compiles.